### PR TITLE
Feature/byoi

### PR DESCRIPTION
**Pull request by guperrot**
Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Shared branch for BYOID.

### CODE REVIEW

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/JwtClaims.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * @param jwt Token in JWT format.
+     * @return Claims or null if parsing failed.
+     */
+    public static JwtClaims parse(String jwt) {
+        String[] parts = jwt.split(JWT_PARTS_SEPARATOR_REGEX);
+        if (parts.length < 2) {
+            AppCenterLog.error(LOG_TAG, "Failed to parse JWT, not enough parts.");
+            return null;
+        }
+        String base64ClaimsPart = parts[1];
```
Comment (achocron) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/JwtClaims.java]: Can you just put a comment explaining what `parts[1]` is?

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 2]
```diff
@@ -447,6 +454,17 @@ public static void startFromLibrary(Context context, Class<? extends AppCenterSe
         return getInstance().setInstanceMaxStorageSizeAsync(storageSizeInBytes);
     }
 
+    /**
+     * Set an authentication provider to refresh authentication tokens when they are about to expire or are already expired.
+     *
+     * @param authProvider authentication provider defined by the application or null to remove it.
+     */
+    @SuppressWarnings("WeakerAccess")
+    // TODO Remove when used in the app without reflection (after release).
```
Comment (jaelim-ms) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: This TODO is for `SuppressWarnings`, please move it above `SuppressWarnings`. I understood it to remove the API after release.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 3]
```diff
@@ -1153,6 +1171,41 @@ public void run() {
...
+            authTokenContext.doNotResetAuthAfterStart();
+            mAuthTokenRefreshListener = new AbstractTokenContextListener() {
+
+                @Override
+                public void onTokenRequiresRefresh(String homeAccountId) {
+                    authProvider.acquireToken(new AuthProvider.Callback() {
+
+                        @Override
+                        public void onAuthResult(String jwt) {
+                            JwtClaims claims = JwtClaims.parse(jwt);
```
Comment (jaelim-ms) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: `parse` doesn't protect `null` JWT. `null` check is required.

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/utils/JwtClaimsTest.java [hunk 4]
```diff
@@ -81,4 +81,11 @@ public void missingParts() {
         JwtClaims jwtClaims = JwtClaims.parse(token);
         assertNull(jwtClaims);
     }
+
+    @Test
+    public void nullToken() {
+        String token = null;
+        JwtClaims jwtClaims = JwtClaims.parse(token);
+        assertNull(jwtClaims);
```
Comment (guperrot) [inline sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/utils/JwtClaimsTest.java]: Might as well inline everything here.

Comment (guperrot) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/utils/JwtClaimsTest.java]: Ok I guess it's to keep the style of the above test.

**Comment by guperrot** [comment]
Data offline still broken for list/delete/write. The last commit fixes only read.

**Pull request closed without merging**